File: Mobzy-API/src/main/java/com/offz/spigot/mobzy/Pathfinders/Flying/PathfinderGoalIdleFly.java
package com.offz.spigot.mobzy.Pathfinders.Flying;

import com.offz.spigot.mobzy.Mobs.Types.FlyingMob;
import net.minecraft.server.v1_13_R2.ControllerMove;
import net.minecraft.server.v1_13_R2.PathfinderGoal;

import java.util.Random;

/**
 * Most code from EntityGhast's pathfinders
 */
public class PathfinderGoalIdleFly extends PathfinderGoal {
    protected final FlyingMob mob;

    public PathfinderGoalIdleFly(FlyingMob flyingMob) {
        this.mob = flyingMob;
    }

    public boolean a() {
        ControllerMove controllermove = mob.getControllerMove();
        if (!controllermove.b()) {
            return true;
        } else {
            double d0 = controllermove.d() - mob.locX;
            double d1 = controllermove.e() - mob.locY;
            double d2 = controllermove.f() - mob.locZ;
            double d3 = d0 * d0 + d1 * d1 + d2 * d2;
            return d3 < 1.0D || d3 > 3600.0D;
        }
    }

    public boolean b() {
        return false;
    }

    public void c() {
        Random random = this.mob.getRandom();
        double x = mob.locX + (double) ((random.nextFloat() * 2.0F - 1.0F) * 16.0F);
        double y = mob.locY + (double) ((random.nextFloat() * 2.0F - 1.5F) * 16.0F); //make it more likely to fly down
        double z = mob.locZ + (double) ((random.nextFloat() * 2.0F - 1.0F) * 16.0F);
        if (y > 16) { //keep mobs from going down and killing themselves
            this.mob.getControllerMove().a(x, y, z, 1.0D);
            mob.lookAt(x, y, z);
        }
    }
}

File: src/main/java/com/mineinabyss/mobzy/pathfinders/flying/PathfinderGoalIdleFly.kt
package com.mineinabyss.mobzy.pathfinders.flying

import com.mineinabyss.mobzy.mobs.types.FlyingMob
import com.mineinabyss.mobzy.pathfinders.MobzyPathfinderGoal
import org.bukkit.Location
import org.bukkit.util.Vector
import kotlin.random.Random

open class PathfinderGoalIdleFly(override val mob: FlyingMob) : MobzyPathfinderGoal() {
    protected var targetLoc: Location? = null

    //if there isn't an operation to move somewhere, we can start looking for somewhere to fly
    override fun shouldExecute(): Boolean = target == null

    override fun shouldKeepExecuting(): Boolean {
        val targetLoc = targetLoc ?: return false
        val dist = mob.distanceTo(targetLoc)
        return dist in 1.0..60.0 && entity.velocity != Vector(0, 0, 0) && target == null
    }

    override fun init() {
        val x = mob.locX + Random.nextDouble(-16.0, 16.0)
        val y = mob.locY + Random.nextDouble(-16.0, 12.0) //make it more likely to fly down
        val z = mob.locZ + Random.nextDouble(-16.0, 16.0)
        val loc = Location(entity.world, x, y, z)
        if (!loc.block.isPassable)
            return
        if (y > 16) { //keep mobs from going down and killing themselves
            targetLoc = loc
            moveController.a(x, y, z, 1.0) //TODO make a wrapper for the controller and figure out the difference between it and navigation
        }
    }

    override fun execute() {
        val targetLoc = targetLoc ?: return
        mob.lookAtPitchLock(targetLoc)
    }
}